tests/unit/test_avrotools.py

```diff
@@ -0,0 +1,146 @@
+"""
+Tests the avrotools module
+"""
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import inspect
+import unittest
+
+import ga4gh.avrotools as avrotools
+import ga4gh.protocol as protocol
+
+
+class TestValidationTool(unittest.TestCase):
+    """
+    Tests the ValidationTool
+    """
+    protocolElementSubclassesNonDefault = set([
+        protocol.GACallSet,
+        protocol.GACigarUnit,
+        protocol.GADataset,
+        protocol.GAException,
+        protocol.GALinearAlignment,
+        protocol.GAListReferenceBasesResponse,
+        protocol.GAPosition,
+        protocol.GAReadAlignment,
+        protocol.GAReadGroup,
+        protocol.GAReadGroupSet,
+        protocol.GAReference,
+        protocol.GAReferenceSet,
+        protocol.GASearchVariantsRequest,
+        protocol.GAVariant,
+        protocol.GAVariantSet,
+        protocol.GAVariantSetMetadata,
+    ])
+
+    def getProtocolElementSubclasses(self):
+        subclasses = []
+        for name, class_ in inspect.getmembers(protocol):
+            if (inspect.isclass(class_) and
+                    issubclass(class_, protocol.ProtocolElement) and
+                    class_ is not protocol.ProtocolElement):
+                subclasses.append(class_)
+        return subclasses
+
+    def testNonProtocolElement(self):
+        """
+        Throws an exception when class_ is not a subclass of ProtocolElement
+        """
+        with self.assertRaises(avrotools.ValidationToolException):
+            avrotools.ValidationTool.getInvalidFields(object, {})
+
+    def testLessFields(self):
+        """
+        Throws an exception when there are fields missing from the jsonDict
+        """
+        with self.assertRaises(avrotools.ValidationToolException):
+            avrotools.ValidationTool.getInvalidFields(protocol.GACall, {})
+
+    def testMoreFields(self):
+        """
+        Throws an exception when there are extra fields in the jsonDict
+        """
+        jsonDict = protocol.GACall().toJsonDict()
+        jsonDict['extra'] = 'extra'
+        with self.assertRaises(avrotools.ValidationToolException):
+            avrotools.ValidationTool.getInvalidFields(
+                protocol.GACall, jsonDict)
+
+    def testDefaultProtocolObjects(self):
+        """
+        Test that any protocol object that can pass validation with
+        a default construction does so
+        """
+        for class_ in self.getProtocolElementSubclasses():
+            if class_ not in self.protocolElementSubclassesNonDefault:
+                self._assertGetInvalidFieldsResultEmpty(class_)
+
+    def testNonDefaultProtocolObjects(self):
+        """
+        Test that any protocol object that can not pass validation with
+        a default construction does so with additional help
+        """
```

Comment:
Can we avoid the hard-coded fields here by instantiating a valid element using `tests.utils.InstanceGenerator`? We should then be able to loop over all classes automatically. Something like:
```python
for class_ in protocol.getProtocolClasses():
    instance = class()
    if not class_.validate(instance.toJsonDict()):
        validInstance = utils.InstanceGenerator()
        # Get the diff between the two dicts and use these to check your method
```
I'd like to define `protocol.getProtocolClasses()` and start using this whenever we're looping over all the subclasses of ProtocolElement. We could do this for all the tests above too, instead of just using `GACall`.